이벤트 페이지 구현

juhwankim-dev opened the pull request:
1. 실패 케이스 부분들은 기획적으로 협의된 부분이 없기 때문에 TODO 달아놓고 비워 놓았음
2. 경품 목록 api response가 바뀌었기 때문에 준혁이가 맡은 쪽에서도 대응해야 할 부분이 있을수도? 확인 부탁!
   - 최대한 내가 짠 코드가 준혁이 코드에 영향 안가도록 짰음. 아마 영향 없을거임!
3. 이벤트 페이지로 최초 진입할때는 문제가 없는데 두번째 이후 진입할 때 리스트를 못가져오는 문제가 있음. 아직 이건 파악중
4. 최초 카카오 로그인하면 응모권 5개 주는 동작을 해야하는데 아직 이거 서버 쪽에서 api 작업 중이라. 마무리 되는대로 이어서 작업하겠음
5. 이벤트 페이지에서 응모를 하면 응모권 개수가 줄어듬. 그리고 뒤로가기하면 홈에서 "내가 가지고 있는 응모권 수"가 줄어들어야 하는데
이에 대해 아직 우리가 논의한게 없어서 동작하지 않을거임
정석은 이벤트 페이지에서 이벤트 쏴주면 홈에서 받아서 반영하는건데
시간이 촉박하다보니 그냥 홈에서 onResume 때마다 티켓수 업데이트 시키는거 어떤지?

juhwankim-dev committed:
feat: 상품 정보를 가져오는 api 구현

juhwankim-dev committed:
chore: 필요한 Resource 추가

juhwankim-dev committed:
feat: 경품 응모 이벤트 ui 구현

juhwankim-dev committed:
feat: 경품 목록 조회 모델 구현

juhwankim-dev committed:
feat: mock 데이터로 화면 구현

juhwankim-dev committed:
feat: 1초마다 공지사항에 보여줄 텍스트(시간)를 viewModel에서 계산한다.

이를 fragment에서 받아 화면에 보여주기 위해 presentation model을 따로 만들었음

juhwankim-dev committed:
feat: 남은 이벤트 기간에 맞게 공지사항을 셋팅하는 로직 구현

marquee로 텍스트가 흐르는 효과도 같이 구현했음

juhwankim-dev committed:
feat: 응모권 받았을 때 뜨는 다이얼로그 구현

juhwankim-dev committed:
feat: 응모권 받았을 때 뜨는 다이얼로그와 이동하는 액션을 네비게이션에 선언

juhwankim-dev committed:
feat: 가지고 있는 응모권 개수가 부족할 때 뜨는 뷰와 로직 구현

juhwankim-dev committed:
feat: 가지고 있는 응모권 개수가 부족한지 여부에 따라 버튼 enabled 수정

juhwankim-dev committed:
feat: 이벤트가 종료되었을 때 이벤트 발표 날짜를 알려주는 뷰와 로직 구현

juhwankim-dev committed:
feat: 이벤트가 종료되었을 때는 가지고 있는 응모권 개수가 부족할 때 뜨는 뷰 gone 처리

juhwankim-dev committed:
feat: 당첨 발표 전과 후의 아이템 분리

juhwankim-dev committed:
feat: item_prize -> item_prize_before_winning_draw 리네이밍

juhwankim-dev committed:
feat: 경품 아이템이 무한하게 반복되도록 구현

juhwankim-dev committed:
feat: 응모하기 클릭시 로티 애니메이션(파티클) 효과를 보여준다.

juhwankim-dev committed:
feat: 응모하기 클릭시 진동 효과를 준다.

juhwankim-dev committed:
feat: 응모 완료시 내 응모권 개수, 응모 횟수, 응모하기 버튼 상태를 최신화한다.

juhwankim-dev committed:
feat: 당첨 발표 후에는 버튼 텍스트가 다르게 뜨게하고, 클릭할 때 다른 동작을 하도록 수정

juhwankim-dev committed:
feat: 당첨 축하 바텀시트 구현

juhwankim-dev committed:
feat: 번호를 다 입력하면 버튼이 활성화 되도록 구현

추가로 13자 까지만 입력이 가능하게 하였음

juhwankim-dev committed:
feat: 번호를 입력하면 자동으로 하이픈이 들어가도록 textWatcher 추가

juhwankim-dev committed:
feat: 연락처 입력 바텀시트에서 연락처를 입력하고 버튼을 눌렀을 때 플로우 구현

닫기 눌렀을때도 같이 구현함~

juhwankim-dev committed:
Merge branch 'feat-115/reward' of https://github.com/Nexters/daily-phrase-android into feature/event

# Conflicts:
#	presentation/src/main/res/navigation/main_navigation.xml
#	presentation/src/main/res/values/colors.xml
#	presentation/src/main/res/values/strings.xml

juhwankim-dev committed:
feat: Domain에 만들어두었던 PrizeInfo 삭제

프로젝트 초기에 만들어두었던 PrizeInfo를 삭제한다.
(현재 api 상황에 맞는 모델을 만들었기 때문)

juhwankim-dev committed:
feat: prizeInfo -> eventInfo로 네이밍 변경

prizeInfo + 이벤트 기간 정보 둘이 합쳐서 eventInfo로 만들거임.

juhwankim-dev committed:
feat: Item -> Prize 네이밍 변경

원래는 PrizeInfo 안에 있는 리스트라 item이라는 네이밍을 사용했으나
PrizeInfo가 EventInfo로 네이밍이 변경되어 수정함

juhwankim-dev committed:
feat: 뷰모델에서 상품, 이벤트, 타이머 정보를 합쳐서 fragment로 방출한다.

juhwankim-dev committed:
feat: flag 값에 따라 adapter에 item을 셋팅한다.

diffUtil을 쓰기엔 무한 스크롤 구현 때문에 비교 연산이 불필요하게 너무 많이 들어가고
매번 리스트를 지우고 새로 넣기에는 1초가 지날때마다 리스트를 새로 넣어야해서
flag값을 두어 아이템을 셋팅하는 방식으로 수정함

juhwankim-dev committed:
feat: 응모 횟수와 응모에 필요한 응모권 갯수 업데이트 로직 추가

뷰페이저 이동할때만 업데이트하고 있었는데
페이지 처음 진입 했을 때 반영이 안돼서 UpdateUi에도 추가함.

juhwankim-dev committed:
feat: 경품 응모 api 구조 구현

juhwankim-dev committed:
feat: 경품 응모 기능 구현

중복 클릭을 막기 위해 로딩도 같이 넣었습니다.

juhwankim-dev committed:
feat: 경품 응모 결과 확인 처리 api 구조 구현

juhwankim-dev committed:
feat: 경품 응모 결과 확인 처리 기능 구현, 결과 확인 했을 때 당첨일 경우 처리

juhwankim-dev committed:
feat: 상태에 따라 응모 관련 뷰 보여주는 로직 구현

juhwankim-dev committed:
feat: ballon 라이브러리 추가

juhwankim-dev committed:
feat: 응모하기 버튼 클릭시 툴팁이 뜨도록 구현

juhwankim-dev committed:
feat: 이벤트 페이지 뒤로가기 버튼 구현

juhwankim-dev committed:
feat: 홈에서 이벤트 페이지로 이동하는 로직 구현

kakaoLogin()에 콜백 함수를 넣지 않은 이유는
이미 기존에도 interface를 만들어서 콜백을 받고 있는데
또 다른 콜백을 추가하면 관리 포인트가 2개로 나뉘어서임.

juhwankim-dev committed:
feat: 당첨자 연락처 입력 api 구조 구현

juhwankim-dev committed:
feat: 당첨자 연락처 입력 로직 구현

juhwankim-dev committed:
feat: 당첨 결과 확인하면 checked를 true로 변경하는 로직 구현

juhwankim-dev committed:
feat: 코드 정리

해당 코드 부분이 무슨 동작을 하는건지 읽는 사람이 쉽게 파악할 수 있도록 중첩 함수로 구현함

nohjunh committed:
feat : 응모 종료 시에 다른 팝업이 뜰 수 있도록 처리

nohjunh committed:
feat : 이벤트가 종료됐을 떄 리워드배너를 띄워주기

nohjunh committed:
feat : 로그인 후 당첨자 안내

nohjunh committed:
feat : 조건처리 명확하게

nohjunh committed:
fix : rewardBanner를 디폴트를 넣지않도록 수정

nohjunh committed:
refactor : 애매한 부분 다시 되돌리기

nohjunh committed:
feat : 응모권 없는 상태에서 10회 이상에 따른 팝업멘트 분기

nohjunh committed:
feat : 가지고 있는 티켓이 있는지 없는지에 따라 보더 색상 다르게

nohjunh committed:
refactor : 동일한 조건식 하나로 묶음

juhwankim-dev committed:
feat: 당첨 발표일 하드코딩 해놨던거 수정

juhwankim-dev committed:
feat: 전화번호 입력시 텍스트가 아니라 editText 객체를 가져왔던 부분 수정

실수..ㅎㅎ

juhwankim-dev committed:
feat: 전화번호 유효성 검사 로직 추가

juhwankim-dev committed:
feat: 상품 아이템 ui 수정

디테일 놓친 부분 보완

juhwankim-dev committed:
fix: 이벤트 페이지에 재진입시 리스트가 빈 상태로 setList하는 버그 수정

viewpager 셋팅하는 시점을 변경했습니다.

nohjunh committed:
feat : 웰컴모달 추가

nohjunh committed:
feat : 응모권 받았어요 툴팁 개선

nohjunh commented on domain/src/main/kotlin/com/silvertown/android/dailyphrase/domain/model/Result.kt:
```diff
@@ -20,6 +20,23 @@ suspend fun <T> Result<T>.onSuccess(
+    }
+}
+
+suspend fun <T> Result<T>.getOrThrow(): T {
```
여기 suspend가 쓰이는 곳이 없는듯!?

juhwankim-dev replied:
[3d21e14](https://github.com/Nexters/daily-phrase-android/pull/116/commits/3d21e14a057a828839c9b109ee4f0fbe389ed8cd) 수정완!

nohjunh commented on domain/src/main/kotlin/com/silvertown/android/dailyphrase/domain/repository/RewardRepository.kt:
```diff
@@ -1,10 +1,16 @@
 
 interface RewardRepository {
     fun getHomeRewardBanner(): Flow<RewardBanner>
+    suspend fun getPrizeInfo(): Result<PrizeInfo> // TODO JH: 나중에 reward든 prize든 통일을 해야할 것 같음
```
동의!

nohjunh commented on presentation/src/main/kotlin/com/silvertown/android/dailyphrase/presentation/ui/event/EventFragment.kt:
```diff
@@ -0,0 +1,245 @@
+import kotlinx.coroutines.launch
+
+@AndroidEntryPoint
+class EventFragment : BaseFragment<FragmentEventBinding>(FragmentEventBinding::inflate) {
```
접근제한자 넣어줘도 될듯

juhwankim-dev replied:
요거 이해를 잘 못하겠는디 어떻게 ??

nohjunh commented on presentation/src/main/kotlin/com/silvertown/android/dailyphrase/presentation/ui/event/EventFragment.kt:
```diff
@@ -0,0 +1,245 @@
+
+        initListeners()
+        initObserve()
+//        showTicketReceivedDialog() // TODO JH: 조건에 따라 보여주기
```
요건 일부러 주석해놓은건가용?

juhwankim-dev replied:
응응 첫 로그인 하면 5장 주는 api가 아직 완성 안되어서 주석해놨구
이후에 수정했습니다여 ~!

nohjunh committed:
Merge branch 'feature/event' into feat/home-reward-detail

nohjunh commented on presentation/src/main/kotlin/com/silvertown/android/dailyphrase/presentation/ui/event/EventFragment.kt:
```diff
@@ -0,0 +1,245 @@
+//        showTicketReceivedDialog() // TODO JH: 조건에 따라 보여주기
+    }
+
+    private fun initListeners() {
+        fun getCurrentPrize(): EventInfoUi.Prize? {
+            return (viewModel.uiState.value as? EventViewModel.UiState.Loaded)
+                ?.eventInfo
+                ?.prizes
+                ?.let { prizes -> prizes[binding.vpPrize.currentItem % prizes.size] }
+        }
+
+        binding.vpPrize.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
+            override fun onPageSelected(position: Int) {
+                super.onPageSelected(position)
+
+                (viewModel.uiState.value as? EventViewModel.UiState.Loaded)?.let {
+                    updateEntryUi(
+                        prize = it.eventInfo.prizes[position % it.eventInfo.prizes.size],
+                        total = it.eventInfo.total,
+                    )
+                }
+            }
+        })
```
가능한 케이스가 있는지는 코드만 봐서는 모르겠지만, 여기 %연산을 할 때 prizes?.takeIf { it.isNotEmpty() } 처럼 빈리스트가 들어오는 경우를 고려해야 exception이 안 뜰 것 같은데 어떻게 보시나요!?

juhwankim-dev replied:
페이지가 선택되었다는 게 리스트가 이미 있다는 전제일 것 같긴 한데
혹시 모르니 일단 감싸놨어! [ceb1de2](https://github.com/Nexters/daily-phrase-android/pull/116/commits/ceb1de28b7378fb5f4f5923f8906395be6de4ac0)

nohjunh commented on presentation/src/main/kotlin/com/silvertown/android/dailyphrase/presentation/ui/event/EventFragment.kt:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    private fun updateUi(eventInfo: EventInfoUi) {
+        if (prizeAdapter == null && eventInfo.prizes.isNotEmpty()) {
+            setViewPager()
+            prizeAdapter?.setList(eventInfo.prizes)
+        }
+        updateEntryUi(
+            prize = eventInfo.prizes[binding.vpPrize.currentItem % eventInfo.prizes.size],
+            total = eventInfo.total,
```
위에서는 isNotEmpty 케이스를 고려하고 있는데, 아래에서는 size를 그대로 받고 있어서 이것도 한 번 확인 부탁!

juhwankim-dev replied:
수정 완! [14c1f95](https://github.com/Nexters/daily-phrase-android/pull/116/commits/14c1f95905c812296bcc624a5d876c732c8db8db)

nohjunh commented on presentation/src/main/kotlin/com/silvertown/android/dailyphrase/presentation/ui/event/EventFragment.kt:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    // 응모와 관련된 ui 업데이트
+    private fun updateEntryUi(prize: EventInfoUi.Prize, total: Int) {
+        binding.tvSubmitEntries.isEnabled = when (prize) {
+            is EventInfoUi.Prize.AfterWinningDraw -> !prize.entryResult.isChecked
+            is EventInfoUi.Prize.BeforeWinningDraw -> !prize.isEventPeriodEnded && prize.hasEnoughEntry
+        }
+        binding.tvMyEntries.text = when (prize) {
+            is EventInfoUi.Prize.AfterWinningDraw -> if (prize.entryResult.isChecked) {
+                when (prize.entryResult.status) {
+                    PrizeInfo.Item.EntryResult.Status.WINNING -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+                        Html.fromHtml(getString(R.string.entry_result_winning, prize.entryResult.phoneNumber), Html.FROM_HTML_MODE_LEGACY)
+                    } else {
+                        Html.fromHtml(getString(R.string.entry_result_winning, prize.entryResult.phoneNumber))
+                    }
+
+                    PrizeInfo.Item.EntryResult.Status.MISSED -> getString(R.string.entry_result_missed)
+                    else -> null
+                }
+            } else {
+                null
+            }
+            is EventInfoUi.Prize.BeforeWinningDraw -> if (prize.isEventPeriodEnded) {
+                null
+            } else {
+                getString(R.string.my_entries, total)
+            }
+        }
+        binding.tvEntryCount.text = getString(R.string.entry_count_message, prize.myEntryCount)
+        binding.tvSubmitEntries.text = if (prize is EventInfoUi.Prize.BeforeWinningDraw) {
+            getString(R.string.submit_entries, prize.requiredTicketCount)
+        } else {
+            getString(R.string.confirm_entry_result)
+        }
+    }
```
여기 binding이 계속 쓰이는 부분은 스코프로 묶어도 될 듯?

juhwankim-dev replied:
요건 취향의 영역일 것 같긴한데
뎁스가 좀 깊어진다 싶어서 안썼어!

nohjunh commented on presentation/src/main/kotlin/com/silvertown/android/dailyphrase/presentation/ui/event/PrizeAdapter.kt:
```diff
@@ -0,0 +1,99 @@
+            binding.tvNotEnoughEntryGuide.isVisible = !prize.hasEnoughEntry && !prize.isEventPeriodEnded
+            binding.tvEventResultReleaseDateGuide.text = itemView.context.getString(
+                R.string.event_result_release_date_guide,
+                String.format("%s월 %s일", prize.winningResultDate.monthValue, prize.winningResultDate.dayOfMonth)
```
요건 따로 strings.xml로 안 빼도 상관없나!?

juhwankim-dev replied:
수정 완! [2b2481e](https://github.com/Nexters/daily-phrase-android/pull/116/commits/2b2481e80855054ae44bda7e50a1678bb37b86df)

nohjunh commented on presentation/src/main/kotlin/com/silvertown/android/dailyphrase/presentation/ui/event/EventViewModel.kt:
```diff
@@ -0,0 +1,312 @@
+
+    fun enterPhoneNumber(prizeId: Int, phoneNumber: String) {
+        suspend fun setPhoneNumberForPrize(prizeId: Int) {
+            (_prizeInfo.value as? Result.Success)?.data?.let { prizeInfo ->
+                prizeInfo.items.map { item ->
+                    if (item.prizeId == prizeId) {
+                        item.copy(entryResult = item.entryResult.toWinningEntryResult(phoneNumber))
+                    } else {
+                        item
+                    }
+                }.let { items ->
+                    prizeInfo.copy(items = items)
+                }.also { prizeInfo ->
+                    _prizeInfo.emit(Result.Success(prizeInfo))
+                }
```
여기 prizeInfo가 nameShadow인 것 같네 실수할 수도 있으니 바꿔처리하는게 좋을듯!

juhwankim-dev replied:
수정 완! [eba4c6a](https://github.com/Nexters/daily-phrase-android/pull/116/commits/eba4c6a670ccd5b8fc2599b0b6f4c018eb7d14d7)

nohjunh committed:
feat : Empty케이스 분기

nohjunh committed:
feat : 웰컴모달 무한스크롤

juhwankim-dev committed:
feat: 티켓 5장을 받을 대상인지 확인하고 팝업 띄워주는 기능 구현

juhwankim-dev committed:
feat: 미당첨인 경우만 응모 결과 확인하기 클릭시 서버 api 호출로 변경

당첨인 경우는 연락처 입력하면 그때 호출하는 것으로 변경 예정

juhwankim-dev committed:
refactor: 코드 정리

checkEntryResult 라고 해놓고서 when문 조건에 따라 checkEntryResult를 안하는 로직이 이상하다고 판단.
fragment에서 호출할 때 애초에 when문으로 분기처리하는 방법으로 변경함

juhwankim-dev committed:
feat: 연락처 입력시 결과 확인했다는 api 호출

juhwankim-dev committed:
feat: 응모하기 api response 수정

Int 형으로 만들어놨네... string이 맞습니다.

nohjunh committed:
refactor : total -> myTicketCount로 변환

nohjunh committed:
feat : 툴팁 이슈 개선

nohjunh committed:
feat : 접근제한자 추가

nohjunh committed:
feat : 합산이 아니라 첫 번째 값으로 처리

nohjunh committed:
feat : eventMonth 필드 대응

nohjunh committed:
feat : 홈 리워드 배너 이미지 고정값으로 하고 패딩은 각 위치별로 따로따로 지정

nohjunh committed:
feat : 딜레이 1000 -> 3000로 변경

juhwankim-dev committed:
fix: 오타 수정

juhwankim-dev committed:
fix: 불필요한 suspend 삭제

juhwankim-dev commented on data/src/main/kotlin/com/silvertown/android/dailyphrase/data/network/model/response/RewardResponse.kt:
```diff
@@ -26,7 +27,18 @@ data class RewardResponse(
+        @SerializedName("phoneNumber")
+        val phoneNumber: String?,
+        @SerializedName("isChecked")
+        val icChecked: Boolean,
```
오우 세상에 [cb2bc39](https://github.com/Nexters/daily-phrase-android/pull/116/commits/cb2bc39f2e971d1d9fbaa7b3788bb84b1e72a345)

juhwankim-dev committed:
fix: onPageSelected 안의 로직을 try catch로 감싼다.

juhwankim-dev committed:
fix: % 연산 사용하는 로직을 try catch로 감싼다.

juhwankim-dev committed:
fix: 하드코딩된 문구를 strings.xml에 추가한다.

juhwankim-dev committed:
fix: 변수명 수정

juhwankim-dev committed:
fix: 문구 수정

nohjunh committed:
feat : ConfirmEntryResultButton도 간격 맞춰주기

nohjunh committed:
feat : welcomeImageUrl, bannerImageUrl 대응

nohjunh committed:
feat : 웰컴모달 자동 스와이프

nohjunh committed:
Merge branch 'feature/event' into feat/home-reward-detail

nohjunh committed:
Merge pull request #117 from Nexters/feat/home-reward-detail

홈 리워드 개선

juhwankim-dev committed:
fix: 다이얼로그에 배경 제대로 적용되지 않는 버그 수정

juhwankim-dev committed:
chore: 불필요한 코드 삭제

juhwankim-dev committed:
fix: api response 변경사항 적용

total -> myTicketCount로 변경된 것 적용함
